haddock-api/src/Haddock.hs

```diff
@@ -247,42 +252,67 @@ withGhc flags action = do
 
 
 readPackagesAndProcessModules :: [Flag] -> [String]
-                              -> Ghc ([(DocPaths, InterfaceFile)], [Interface], LinkEnv)
+                              -> Ghc ([(DocPaths, FilePath, InterfaceFile)], [Interface], LinkEnv)
 readPackagesAndProcessModules flags files = do
     -- Get packages supplied with --read-interface.
     let noChecks = Flag_BypassInterfaceVersonCheck `elem` flags
     packages <- readInterfaceFiles nameCacheFromGhc (readIfaceArgs flags) noChecks
 
     -- Create the interfaces -- this is the core part of Haddock.
-    let ifaceFiles = map snd packages
+    let ifaceFiles = map (\(_, _, ifaceFile) -> ifaceFile) packages
     (ifaces, homeLinks) <- processModules (verbosity flags) files flags ifaceFiles
 
     return (packages, ifaces, homeLinks)
 
 
 renderStep :: DynFlags -> [Flag] -> SinceQual -> QualOption
-           -> [(DocPaths, InterfaceFile)] -> [Interface] -> IO ()
+           -> [(DocPaths, FilePath, InterfaceFile)] -> [Interface] -> IO ()
 renderStep dflags flags sinceQual nameQual pkgs interfaces = do
-  updateHTMLXRefs pkgs
+  updateHTMLXRefs (map (\(docPath, _ifaceFilePath, ifaceFile) ->
+                          ( case baseUrl flags of
+                              Nothing  -> fst docPath
+                              Just url -> url </> packageName (ifUnitId ifaceFile)
+                          , ifaceFile)) pkgs)
```

Comment:
@Kleidukos this the crux of the fix for local link. Ideally we would allow the user to provide format for a path that would be constructed from package name, version etc.  This is something much simpler, just `baseUrl </> packageName`.  We could also use `show (ifUnitId ifaceFile)`, but the drawback is that `UnitId` have version and `-inplace` suffix as well (for local packages). With a bit of simple string manipulation I can get package name (the `packageName` function), I couldn't find a function that extracts package name from `Unit` (if you know about anything like that somewhere in `ghc`, I'd appreciate). Altogether this works fine, but it forces how the target directory looks like.  Using format string would lift this assumption, but I am not sure if this is worth the effort (at least in the initial version).

Reply:
Now all links here work fine: https://coot.github.io/haddock-example/package-b/B.html

This also shows that one can include upstream packages, in this case `base` package. Using quick-search for identifiers in `base` also works :).